File: funkotlin/src/main/kotlin/funkotlin/Try.kt
package funkotlin

sealed class Try<out T> {

    data class Success<T>(val value: T) : Try<T>()
    data class Failure(val throwable: Throwable) : Try<Nothing>()

    fun <R> fold(g: (Throwable) -> R, f: (T) -> R): R = when (this) {
        is Success -> f(value)
        is Failure -> g(throwable)
    }

    fun <R> flatMap(f: (T) -> Try<R>): Try<R> = fold(
            { this as Failure },
            { f(it) }
    )

    fun <R> map(f: (T) -> R): Try<R> = flatMap { Success(f(it)) }

    fun foreach(f: (T) -> Unit): Unit = fold({}, f)

    fun exists(f: (T) -> Boolean): Boolean = fold(
            { false },
            { f(it) }
    )

    val isSuccess = fold(
            { false },
            { true }
    )

    val isFailed = !isSuccess

    fun toOption(): Option<T> = fold(
            { Option.None },
            { Option.Some(it) }
    )

}

fun <T> Try<T>.filter(f: (T) -> Boolean): Try<T> = fold(
        { this },
        {
            val value: Try<T> = try {
                if (f(it)) {
                    this as Try.Success<T>
                } else {
                    Try.Failure(NoSuchElementException("Predicate does not hold for $it"))
                }
            } catch (t: Throwable) {
                Try.Failure(t)
            }
            value
        }
)

fun <T> Try<T>.filterNot(f: (T) -> Boolean): Try<T> = filter { !f(it) }

fun <T> Try<T>.getOrElse(t: () -> T): T = fold(
        { t() },
        { it }
)

fun <T> Try<T>.orElse(f: () -> Try<T>): Try<T> = fold(
        { f() },
        { this }
)

fun <T> Try<T>.recover(f: (Throwable) -> Try<T>): Try<T> = fold(
        { f(it) },
        { this }
)

File: funscala/src/main/scala/funscala/Try.scala
package funscala

sealed trait Try[+T] extends MonadMethod[Try, T] {

  override def identity[TT >: T]: Try[TT] = Failure(new IllegalArgumentException("Element doesn't exist"))

  @inline def fold[R](f: => R, g: T => R): R = this match {
    case Failure(_) => f
    case Success(v) => g(v)
  }

  @inline override def flatMap[R](f: T => Try[R]): Try[R] = this match {
    case f@Failure(_) => f
    case Success(v) => f(v)
  }

  @inline def isSuccess: Boolean = fold(false, _ => true)

  @inline def isFailed: Boolean = !isSuccess

  @inline def toOption: Option[T] = fold(None, Some(_))

  @inline def recover[TT >: T](f: => Try[TT]): Try[TT] = fold(f, _ => this)
  
}

object Try {

  implicit def pureable[T] = new Pureable[Try,T] {
    override def pure[X](x: X): Try[X] = Success(x)
  }

}

final case class Success[T](value: T) extends Try[T]
final case class Failure(throwable: Throwable) extends Try[Nothing]